cpp/include/raft/core/math.hpp

```diff
@@ -100,9 +112,49 @@ RAFT_INLINE_FUNCTION auto cos(T x)
 #endif
 }
 
+#if _RAFT_HAS_CUDA
+template <typename T>
+RAFT_DEVICE_INLINE_FUNCTION typename std::enable_if<std::is_same<T, __half>::value, __half>::type
+cos(T x)
+{
+#if (__CUDA_ARCH__ >= 530)
+  return ::hcos(x);
+#else
+  // static_assert(false) would be evaluated during host compilation stage while __CUDA_ARCH__ is at
+  // device compilation stage Using this sizeof(T) != sizeof(T) makes it work as it's only triggered
+  // during template instantiation and thus at device compilation stage
+  static_assert(sizeof(T) != sizeof(T), "__half is only supported on __CUDA_ARCH__ >= 530");
+  return T{};
+#endif
```

Comment:
Question for @cjnolet : iirc RAPIDS is Pascal+, shall we have a centralized check of the minimum arch somewhere and remove these checks here, or would this have too many consequences on repos using raft?